### PR TITLE
[FLINK-8451] [serializers] Make Scala tuple serializer deserialization more failure tolerant

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

Scala tuples (and thus also case classes) classes are serialized together with their serializers. The `serialVersionUID` changed from Scala 2.10 to 2.11. We need to ignore those changes when reading a serializer config snapshot.

## Brief change log

- Make deserialization more failure tolerant

## Verifying this change

- See `org.apache.flink.api.scala.runtime.TupleSerializerCompatibilityTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by twalthr** [comment]
CC @tzulitai @aljoscha

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 1]
```diff
@@ -121,10 +121,53 @@ public ClassLoaderObjectInputStream(InputStream in, ClassLoader classLoader) thr
...
 		scalaSerializerClassnames.add("org.apache.flink.api.scala.typeutils.UnitSerializer");
 	}
 
+	/**
+	 * The serialVersionUID might change between Scala versions and since those classes are
+	 * part of the tuple serializer config snapshots we need to ignore them.
+	 *
+	 * See also FLINK-8451.
+	 */
+	private static Set<String> scalaTypes = new HashSet<>();
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: make final?

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 2]
```diff
@@ -121,10 +121,53 @@ public ClassLoaderObjectInputStream(InputStream in, ClassLoader classLoader) thr
...
+		scalaTypes.add("scala.Tuple15");
+		scalaTypes.add("scala.Tuple16");
+		scalaTypes.add("scala.Tuple17");
+		scalaTypes.add("scala.Tuple18");
+		scalaTypes.add("scala.Tuple19");
+		scalaTypes.add("scala.Tuple20");
+		scalaTypes.add("scala.Tuple21");
+		scalaTypes.add("scala.Tuple22");
+		scalaTypes.add("scala.Tuple1$mcJ$sp");
+		scalaTypes.add("scala.Tuple1$mcI$sp");
```
Comment (tzulitai) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: Could you briefly explain what these are?

**Comment by tzulitai** [comment]
A few side notes:
1) Maybe we should have written just the arity of tuples in the config snapshots. That would have avoided this issue.
2) Is there any potential cases like this one, for other Scala types beside Tuples?

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java (hunk 2 above)
Comment (twalthr) [reply flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: These anonymous classes are used for type specific tuples. Like `II` for a `(Int, Int)`. I don't know exactly when they are inserted by the compiler but I added them just to be sure.

**Comment by twalthr** [comment]
@tzulitai I think just writing out the arity is not enough. It really depends on the type of tuple. Maybe it would have been better to write out only the name of the class. But since we support tuple POJOs this might be a little bit risky. The tuple serializer serializes general subclasses for both Java and Scala and case classes.

I tested case classes, tuples, `Either` and `Unit`. What do you think we should test as well? The other serializers do not include classes that need to be snapshotted.

**Comment by tzulitai** [comment]
@twalthr I see. Then for other Scala types whose serializer config snapshot does not require writing classes, we probably don't need to include them in the `FailureTolerantObjectInputStream` then.

**Comment by twalthr** [comment]
@tzulitai exactly, I think tuples/case classes are just a special case. @aljoscha are we going to merge this now? (given the checkstyle error is fixed)

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 3]
```diff
@@ -113,18 +113,61 @@ public ClassLoaderObjectInputStream(InputStream in, ClassLoader classLoader) thr
...
-	private static Set<String> scalaSerializerClassnames = new HashSet<>();
+	private final static Set<String> scalaSerializerClassnames = new HashSet<>();
 	static {
 		scalaSerializerClassnames.add("org.apache.flink.api.scala.typeutils.TraversableSerializer");
 		scalaSerializerClassnames.add("org.apache.flink.api.scala.typeutils.CaseClassSerializer");
 		scalaSerializerClassnames.add("org.apache.flink.api.scala.typeutils.EitherSerializer");
 		scalaSerializerClassnames.add("org.apache.flink.api.scala.typeutils.EnumValueSerializer");
 		scalaSerializerClassnames.add("org.apache.flink.api.scala.typeutils.OptionSerializer");
 		scalaSerializerClassnames.add("org.apache.flink.api.scala.typeutils.TrySerializer");
-		scalaSerializerClassnames.add("org.apache.flink.api.scala.typeutils.EitherSerializer");
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: Why is this removed?

Comment (aljoscha) [reply flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: nvm, @twalthr pointed out that it was in there twice

**Comment by StephanEwen** [comment]
+1 to what Gordon suggested.

Let's move towards avoiding writing classes alltogether in any config snapshot. Tuple arity is the right thing for tuples, plus ideally a format version number.

**Comment by twalthr** [comment]
@StephanEwen a tuple arity would be enough if the tuple serializer would only serialize tuples but it also serializes case classes and POJOs. Wouldn't be the fully qualified name better? But yes writing classes should be avoided.
